connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetQueryVisitor.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.teiid.translator.parquet;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Stack;
+
+import org.teiid.language.*;
+import org.teiid.language.visitor.HierarchyVisitor;
+import org.teiid.metadata.Column;
+import org.teiid.metadata.Table;
+import org.teiid.translator.TranslatorException;
+
+public class ParquetQueryVisitor extends HierarchyVisitor {
+
+    protected Stack<Object> onGoingExpression = new Stack<Object>();
+    private List<String> projectedColumnNames = new ArrayList<String>();
+    protected ArrayList<TranslatorException> exceptions = new ArrayList<TranslatorException>();
+
+    public HashMap<String, String> getColumnValues() {
+        return columnValues;
+    }
+
+    private HashMap<String, String> columnValues = new HashMap<>();
+
+    private Table table;
+    private String parquetPath;
+
+    public List<String> getProjectedColumnNames() {
+        return projectedColumnNames;
+    }
+
+    public ArrayList<TranslatorException> getExceptions() {
+        return exceptions;
+    }
+
+    public Table getTable() {
+        return table;
+    }
+
+    public String getParquetPath() {
+        return parquetPath;
+    }
+
+    @Override
+    public void visit(NamedTable obj) {
+        this.table = obj.getMetadataObject();
+        this.parquetPath = this.table.getProperty(ParquetMetadataProcessor.FILE);
+    }
+
+    @Override
+    public void visit(ColumnReference obj) {
+        this.onGoingExpression.add(obj.getMetadataObject());
+    }
+
+    @Override
+    public void visit(DerivedColumn obj) {
+        visitNode(obj.getExpression());
+
+        createProjectedColumn();
+    }
+
+    private void createProjectedColumn() {
+        Column column = (Column) this.onGoingExpression.pop();
+        this.projectedColumnNames.add(column.getSourceName());
+    }
+
+    @Override
+    public void visit(Comparison obj) {
+        if(this.table.getProperty(ParquetMetadataProcessor.PARTITIONING_SCHEME) == null){
+            // TODO: Row filtering
+        }
+        else {
+            if (this.table.getProperty(ParquetMetadataProcessor.PARTITIONING_SCHEME).equals("directory")) {
+                directoryBasedPartitioning(obj);
+            } else {
+                // TODO: File Based Partitioning
+            }
+        }
+    }
+
+    @Override
+    public void visit(Literal obj) {
+        this.onGoingExpression.push(obj.getValue());
+    }
+
+    private void directoryBasedPartitioning(Comparison obj) {
+        visitNode(obj.getLeftExpression());
+        Column column = (Column)this.onGoingExpression.pop();
+        String columnName = column.getSourceName();
+        visitNode(obj.getRightExpression());
+        String columnValue = (String) this.onGoingExpression.pop();
```

Comment:
Thinking beyond the partitioning columns it is not required that the literal value will be a String.  It can be any of the comparable value types that are supported by the translator.  For the partitioning columns we would require that the type value is easily convertable to string - for example exact numeric would be allowable as well.  It's important to have the types represented correctly as there is a semantic difference between string an numeric types that comes into play once you support other comparisons.  For example with 10 > 2 as an integer comparison, but '10' < '2' as a string comparison.

Also, since the visitor is performing casts in any case, you may find it more straight-forward to do something like:

public void visit(Comparison obj) {
  ColumnReference cr = (ColumnReference)obj.getLeftExpression();
  Column column = cr.getMetadataObject();
  Literal l = (Literal)obj.getRightExpression();
  ...
}

The ongoing expression logic will only be helpful if there ends up being handling that is common for very large expressions / subtrees.

If you want to consider the other comparisons and IN predicates at this point you also want to make what you are capturing pretty general - it will need more than just the column and the value to also convey the type of operation (for example less than vs. equality) and a set of values (IN ...)

Reply:
> Thinking beyond the partitioning columns it is not required that the literal value will be a String.

Yes, I initially took it to be just the directory partitioning where the user can be asked to name the column as strings. But since we would also need to handle other aspects and comparison types other than equality, this would fail, we need to know the types of values too. What all types are we going to handle? Also, what datatypes would allow what kind of comparisons is also a question. For eg, I don't think a complex data type like an array would be supported. Isn't it?

> The ongoing expression logic will only be helpful if there ends up being handling that is common for very large expressions / subtrees.

Right! The visiting of them is not needed, I was also wondering if there's a need for this visiting even in the excel query visitor. That also seems similar and could be simplified like you mentioned. 

>If you want to consider the other comparisons and IN predicates at this point you also want to make what you are capturing pretty general - it will need more than just the column and the value to also convey the type of operation (for example less than vs. equality) and a set of values (IN ...)

I think I should rather implement all of it together which would make the capability wholesome and things can be implemented accordingly. Otherwise, like you said they would have to be generalized or changed anyway in the future when the support is extended. What do you say?

Reply:
>  What all types are we going to handle? Also, what datatypes would allow what kind of comparisons is also a question. For eg, I don't think a complex data type like an array would be supported. Isn't it?

Generally the translator will need to support literal values that match the return values to the engine.  In this case you have string, varbinary, biginteger, long, integer, float, double, boolean - all of those are comparable types that can end up in a predicate of the form "col = value literal".  However partitioning columns may need additional restrictions, such as not using varbinary, float, double values - as those could have ambiguous string values.  

> Otherwise, like you said they would have to be generalized or changed anyway in the future when the support is extended. What do you say?

It will be more difficult to fully implement everything at once, but you can make additional logic easier just by capturing more information, rather than less, in your visitation / query analysis.  Some things will be pretty easier to layer in, such as once you support predicates on non-partitioning columns you'll need to add the logic to convert those predicates into a filter.  Other things will be more difficult, such as once you support comparisons other than equality, then you can't simply build up a file pattern to honor the predicate - you'd have to implement a check against the file path - which will also require updating the VirtualFile api to expose both a file name and a file path.

Reply:
> However partitioning columns may need additional restrictions, such as not using varbinary, float, double values - as those could have ambiguous string values.

So, I'm implementing the visitor as something that builds a map on the HashMap<String(ColumnName), Literal>. So when I get the literal corresponding to a column, I would implement it using a switch case using the class name(Is there an enum for the same, I mean an API that would give us the teiid types for a corresponding literal) that would convert the respective string,  biginteger, long, integer,  boolean to strings using the toString methods and default the switch case to throw an exception for incompatible types for a directory path like the varbinary, float, and double. Thoughts?

>  It will be more difficult to fully implement everything at once

Alright, we'll do things one step at a time and make progress that way. 

> you'd have to implement a check against the file path - which will also require updating the VirtualFile api to expose both a file name and a file path.

Right!!

Reply:
> Thoughts?

Yes that is the right track.  Capture the full literal, or even the full predicate, for further work.

Also note that opening up to additional comparisons allows for thing like "between" scenarios - where col > 'value1' and col < 'value2' - which will further complicate handling.

Reply:
> or even the full predicate

Like just store the complete comparison object, and later use it for different purposes, like here we can probably utilize it directly to build the directory path, right?

> Also note that opening up to additional comparisons allows for thing like "between" scenarios - where col > 'value1' and col < 'value2' - which will further complicate handling.

This would be after the virtualfile API is capable to give the file path, right?

Reply:
> Like just store the complete comparison object, and later use it for different purposes, like here we can probably utilize it directly to build the directory path, right?

Yes a map of columns to predicates would be the most general.

> This would be after the virtualfile API is capable to give the file path, right?

Yes, and I'm trying to highlight that there could be more than one predicate for a given column.  That can be handled by a multi-map or similar.